### PR TITLE
Allow RequestID to be shrunk if possible

### PR DESCRIPTION
**Pull request by upsampled**
fixes gosnmp/gosnmp#450  and also unblocks gosnmp/gosnmp#448 (first test passes after this is merged)

### CODE REVIEW

File: marshal.go [hunk 1]
```diff
@@ -1379,3 +1378,22 @@ func (x *GoSNMP) receive() ([]byte, error) {
 	copy(resp, x.rxBuf[:n])
 	return resp, nil
 }
+
+func shrinkAndWriteUint(buf io.Writer, id uint) error {
+	var pktid any
+	switch {
+	case id < math.MaxUint8:
```
Comment (SuperQ) [inline marshal.go:1385]: Should these be `<=`? For example, `255` should fit in 1 byte, but this would make it stored in 2 bytes.

**Review by SuperQ — CHANGES_REQUESTED** [review]
I think we need some unit tests for the new func.

File: marshal.go [hunk 2]
```diff
@@ -1379,3 +1378,22 @@ func (x *GoSNMP) receive() ([]byte, error) {
...
+		buf.Write([]byte{2, 1}) //2 is type 1 is length
```
Comment (SuperQ) [inline marshal.go:1386]: One suggestion, move this write to the end.

Something like this.
```
  var padding int
  case ...:
    padding = 1
  case ...:
    padding = 2
  } 
  buf.Write([]byte{2, padding})
  return ...
```

**Comment by upsampled** [comment]
I ran a quick test and saw that `asn1` handles this way better than I could.  It even handles and odd number of bytes (IE: `math.MaxUint16 + 1` will be encoded to `0203010000` instead of  `020400010000`).

**Review by SuperQ — APPROVED** [review]
Nice

**Review by TimRots — APPROVED** [review]
Nice, thank you

**Pull request merged**